### ISSUE TITLE
Gitlab support

### ISSUE THREAD

**Issue by hipstersmoothie** [intuit/auto-release#203]
**Is your feature request related to a problem? Please describe.**

`auto` currently only works with GitHub. 

**Describe the solution you'd like**

A plugin for GItLab which is loaded automatically for gitlab URLS.

**Describe alternatives you've considered**

**Comment by zephraph** [comment]
Hopefully when #184 is done it'll make this a little easier.

**Comment by hipstersmoothie** [comment]
<img width="905" alt="screen shot 2019-01-22 at 1 38 14 pm" src="https://user-images.githubusercontent.com/1192452/51567036-0c583a80-1e4b-11e9-9e57-7d2c71a16bf4.png">

**Comment by sc250024** [comment]
Can I ask, what's the status on this? @zephraph 's previous comment was on a PR that was closed. Is something needed for this project? Is something missing from the GitLab API?

**Comment by hipstersmoothie** [comment]
I dont see this being all that much work. We have this class https://github.com/intuit/auto/blob/master/packages/core/src/git.ts that handles setting up ocotkit and making calls to github. All we need to do is:

- detect github/gitlab from repo url
- call appropriate url

The hardest part of this might be the types. Octokit has very comprehensive types and we use them in a few places.

**Comment by james-s-nduka** [comment]
Hi - any updates to this? Would really appreciate if this was Gitlab compatible

**Comment by hipstersmoothie** [comment]
It's a pretty large lift without an octokit for GitLab. To make this work we basically need to write that

**Comment by jdalrymple** [comment]
> It's a pretty large lift without an octokit for GitLab. To make this work we basically need to write that

My time to shine hehe, I created a library called [Gitbeaker](https://github.com/jdalrymple/gitbeaker) to do just this!

**Comment by jdalrymple** [comment]
Might be smarter to split this part out into a plugin, thinking about supporting different hosting services in the future.

**Comment by hipstersmoothie** [comment]
> Might be smarter to split this part out into a plugin

Exactly! In an ideal world Github stuff should become a plugin and GitLab will just implement a similar interface.

Getting the types to work nicely might be hard though. there are a few places where plugins an parts of auto access octokit directly too

**Comment by hipstersmoothie** [comment]
@jdalrymple your work looks super promising!

**Comment by hipstersmoothie** [comment]
I think the best way to approach this is to:

1. get some of the lower hanging fruit out of the way (label, create-labels, pr-check, pr-status, pr-body, comment)
2. While doing `1` craft the base of what GitHub being a plugin looks like  so we can then tackle the deeper issues

I think I'll start with `comment`. Can you send me the appropriate APIs to create, edit and delete comments on PRs?

**Comment by jdalrymple** [comment]
> I think the best way to approach this is to:
> 
>     1. get some of the lower hanging fruit out of the way (label, create-labels, pr-check, pr-status, pr-body, comment)
> 
>     2. While doing `1` craft the base of what GitHub being a plugin looks like  so we can then tackle the deeper issues
> 
> 
> I think I'll start with `comment`. Can you send me the appropriate APIs to create, edit and delete comments on PRs?

Heres the [Gitlab API docs](https://docs.gitlab.com/ee/api/notes.html#merge-requests)
And the corresponding [gitbeaker support](https://github.com/jdalrymple/gitbeaker/blob/master/packages/gitbeaker-core/src/services/MergeRequestNotes.ts)

Sorry for the raw source, I'm still working on getting docs up and running

**Comment by lukasmrtvy** [comment]
Any progress on this ? Thanks

**Comment by jdalrymple** [comment]
@lukasmrtvy I dont think the github support was ever moved out into a plugin format (I could be wrong!) but the support definitely exists in Gitbeaker to connect to gitlab.

**Comment by serafdev** [comment]
@jdalrymple it is a package, if someone can propose an interface for it and link to this issue we can work from there: https://github.com/intuit/auto/tree/main/plugins/git-tag
